### ISSUE TITLE
Allow @IF(<iface>) seems not to work as expected

### ISSUE THREAD

**Issue by jirib** [apple/cups#5918]
Hello,

`cups-2.3.3-3.1.x86_64` here on OpenSUSE Tumbleweed. I tried to use `@IF(enp2s0)` in *Location* but it seems not to work as expected, clients can't get printer - forbidden.

```
# head -n 15 /etc/cups/cupsd.conf
LogLevel debug
PageLogFormat
# Allow remote access
Port 631
Listen /run/cups/cups.sock
# Share local printers on the local network.
Browsing On
BrowseLocalProtocols dnssd
DefaultAuthType Basic
WebInterface yes
<Location />
  # Allow shared printing...
  Order allow,deny
  Allow @IF(enp2s0)
</Location>
```
```
# ip a s dev enp2s0
2: enp2s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 00:25:90:60:8f:1e brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.1/24 brd 192.168.1.255 scope global enp2s0
       valid_lft forever preferred_lft forever
    inet6 fe80::225:90ff:fe60:8f1e/64 scope link 
       valid_lft forever preferred_lft forever
```

```
# cupsd -t
I [15/Apr/2021:21:19:22 +0200] Listening to 0.0.0.0:631 (IPv4)
I [15/Apr/2021:21:19:22 +0200] Listening to [v1.::]:631 (IPv6)
I [15/Apr/2021:21:19:22 +0200] Listening to /run/cups/cups.sock (Domain)
I [15/Apr/2021:21:19:22 +0200] Remote access is enabled.
D [15/Apr/2021:21:19:22 +0200] Added auto ServerAlias router
I [15/Apr/2021:21:19:22 +0200] Loaded configuration file "/etc/cups/cupsd.conf"
D [15/Apr/2021:21:19:22 +0200] Using keychain "/etc/cups/ssl" for server name "router".
I [15/Apr/2021:21:19:22 +0200] Using default TempDir of /var/spool/cups/tmp...
I [15/Apr/2021:21:19:22 +0200] Configured for up to 100 clients.
I [15/Apr/2021:21:19:22 +0200] Allowing up to 100 client connections per host.
I [15/Apr/2021:21:19:22 +0200] Full reload is required.
I [15/Apr/2021:21:19:22 +0200] Loaded MIME database from "/usr/share/cups/mime" and "/etc/cups": 78 types, 114 filters...
D [15/Apr/2021:21:19:22 +0200] Loading printer epsonalm2000...
D [15/Apr/2021:21:19:22 +0200] load_ppd: Loading /var/cache/cups/epsonalm2000.data...
D [15/Apr/2021:21:19:22 +0200] cupsdRegisterPrinter(p=0x558afc08b2f0(epsonalm2000))
I [15/Apr/2021:21:19:22 +0200] Loading job cache file "/var/cache/cups/job.cache"...
I [15/Apr/2021:21:19:22 +0200] Full reload complete.
"/etc/cups/cups-files.conf" is OK.
"/etc/cups/cupsd.conf" is OK.
# cupsd -f
...
```

Now from a client:

```
# lpstat -h 192.168.1.1:631 -v
lpstat: Forbidden
```

`cupsd -f` output
```
...
D [15/Apr/2021:21:14:00 +0200] cupsdSetBusyState: newbusy="Active clients", busy="Not busy"
D [15/Apr/2021:21:14:00 +0200] [Client 36] Server address is "192.168.1.1".
D [15/Apr/2021:21:14:00 +0200] [Client 36] Accepted from 192.168.1.194:46990 (IPv4)
D [15/Apr/2021:21:14:00 +0200] [Client 36] Waiting for request.
D [15/Apr/2021:21:14:00 +0200] [Client 36] POST / HTTP/1.1
D [15/Apr/2021:21:14:00 +0200] cupsdSetBusyState: newbusy="Active clients", busy="Active clients"
D [15/Apr/2021:21:14:00 +0200] [Client 36] Read: status=200, state=6
D [15/Apr/2021:21:14:00 +0200] [Client 36] No authentication data provided.
D [15/Apr/2021:21:14:00 +0200] [Client 36] cupsdSendHeader: code=403, type="text/html", auth_type=0
D [15/Apr/2021:21:14:00 +0200] [Client 36] Closing connection.
D [15/Apr/2021:21:14:00 +0200] cupsdSetBusyState: newbusy="Not busy", busy="Active clients"
D [15/Apr/2021:21:14:00 +0200] cupsdSetBusyState: newbusy="Active clients", busy="Not busy"
D [15/Apr/2021:21:14:00 +0200] [Client 37] Server address is "192.168.1.1".
D [15/Apr/2021:21:14:00 +0200] [Client 37] Accepted from 192.168.1.194:46992 (IPv4)
D [15/Apr/2021:21:14:00 +0200] [Client 37] Waiting for request.
D [15/Apr/2021:21:14:00 +0200] [Client 37] HTTP_STATE_WAITING Closing for error 32 (Broken pipe)
D [15/Apr/2021:21:14:00 +0200] [Client 37] Closing connection.
D [15/Apr/2021:21:14:00 +0200] cupsdSetBusyState: newbusy="Not busy", busy="Active clients"
D [15/Apr/2021:21:14:00 +0200] cupsdSetBusyState: newbusy="Active clients", busy="Not busy"
D [15/Apr/2021:21:14:00 +0200] [Client 38] Server address is "192.168.1.1".
D [15/Apr/2021:21:14:00 +0200] [Client 38] Accepted from 192.168.1.194:46994 (IPv4)
D [15/Apr/2021:21:14:00 +0200] [Client 38] Waiting for request.
D [15/Apr/2021:21:14:00 +0200] [Client 38] POST / HTTP/1.1
D [15/Apr/2021:21:14:00 +0200] cupsdSetBusyState: newbusy="Active clients", busy="Active clients"
D [15/Apr/2021:21:14:00 +0200] [Client 38] Read: status=200, state=6
D [15/Apr/2021:21:14:00 +0200] [Client 38] No authentication data provided.
D [15/Apr/2021:21:14:00 +0200] [Client 38] cupsdSendHeader: code=403, type="text/html", auth_type=0
D [15/Apr/2021:21:14:00 +0200] [Client 38] Closing connection.
D [15/Apr/2021:21:14:00 +0200] cupsdSetBusyState: newbusy="Not busy", busy="Active clients"
D [15/Apr/2021:21:14:00 +0200] cupsdSetBusyState: newbusy="Active clients", busy="Not busy"
D [15/Apr/2021:21:14:00 +0200] [Client 39] Server address is "192.168.1.1".
D [15/Apr/2021:21:14:00 +0200] [Client 39] Accepted from 192.168.1.194:46996 (IPv4)
D [15/Apr/2021:21:14:00 +0200] [Client 39] Waiting for request.
D [15/Apr/2021:21:14:00 +0200] [Client 39] HTTP_STATE_WAITING Closing for error 32 (Broken pipe)
D [15/Apr/2021:21:14:00 +0200] [Client 39] Closing connection.
D [15/Apr/2021:21:14:00 +0200] cupsdSetBusyState: newbusy="Not busy", busy="Active clients"
D [15/Apr/2021:21:14:00 +0200] cupsdSetBusyState: newbusy="Active clients", busy="Not busy"
D [15/Apr/2021:21:14:00 +0200] [Client 40] Server address is "192.168.1.1".
D [15/Apr/2021:21:14:00 +0200] [Client 40] Accepted from 192.168.1.194:46998 (IPv4)
D [15/Apr/2021:21:14:00 +0200] [Client 40] Waiting for request.
D [15/Apr/2021:21:14:00 +0200] [Client 40] POST / HTTP/1.1
D [15/Apr/2021:21:14:00 +0200] cupsdSetBusyState: newbusy="Active clients", busy="Active clients"
D [15/Apr/2021:21:14:00 +0200] [Client 40] Read: status=200, state=6
D [15/Apr/2021:21:14:00 +0200] [Client 40] No authentication data provided.
D [15/Apr/2021:21:14:00 +0200] [Client 40] cupsdSendHeader: code=403, type="text/html", auth_type=0
D [15/Apr/2021:21:14:00 +0200] [Client 40] Closing connection.
D [15/Apr/2021:21:14:00 +0200] cupsdSetBusyState: newbusy="Not busy", busy="Active clients"
```

Is `@IF(enp2s0)` wrong or it's a bug?

```
# man cupsd.conf | col -b | grep -A 5 'Allow @IF(name)' | fmt -w 80
       Allow @IF(name)

       Allow @LOCAL
            Allows access from the named hosts, domains, addresses,
            or interfaces.  The @IF(name) form uses the current subnets
            configured for the named interface.  The @LOCAL form uses the
            current subnets configured for  all  inter- faces that are not
            point-to-point, for example Ethernet and Wi-Fi interfaces are used
            but DSL and VPN interfaces are not.  The Order directive controls
            whether Allow lines are evaluated before or after Deny lines.
```

**Comment by michaelrsweet** [comment]
@jirib What kind of interface is "enp2s0"? It is possible cupsd doesn't think it is a valid (non-p2p) interface...

**Comment by jirib** [comment]
It's typical *e1000* interface.

```
# ethtool -i enp2s0
driver: e1000e
version: 5.11.6-1-default
firmware-version: 1.9-0
expansion-rom-version: 
bus-info: 0000:02:00.0
supports-statistics: yes
supports-test: yes
supports-eeprom-access: yes
supports-register-dump: yes
supports-priv-flags: yes

# lspci -v -s 02:00.0
02:00.0 Ethernet controller: Intel Corporation 82574L Gigabit Network Connection
        Subsystem: Super Micro Computer Inc Device 10d3
        Physical Slot: 0-1
        Flags: bus master, fast devsel, latency 0, IRQ 16
        Memory at fe9e0000 (32-bit, non-prefetchable) [size=128K]
        I/O ports at dc00 [size=32]
        Memory at fe9dc000 (32-bit, non-prefetchable) [size=16K]
        Capabilities: [c8] Power Management version 2
        Capabilities: [d0] MSI: Enable- Count=1/1 Maskable- 64bit+
        Capabilities: [e0] Express Endpoint, MSI 00
        Capabilities: [a0] MSI-X: Enable+ Count=5 Masked-
        Capabilities: [100] Advanced Error Reporting
        Capabilities: [140] Device Serial Number 00-25-90-ff-ff-60-8f-1e
        Kernel driver in use: e1000e
        Kernel modules: e1000e
```

**Comment by michaelrsweet** [comment]
@jirib Hmm, OK. Can you enable debug logging (`cupsctl --debug-logging`) and then look for `cupsdNetIFUpdate` messages in the /var/log/cups/error_log file?

**Comment by michaelrsweet** [comment]
@jirib I should also note: `@IF(name)` doesn't detect whether the network traffic came over the named interface. Rather, it expands to allow/deny access to all of that interface's subnets. So in your case `@IF(name)` expands to 192.168.1.0/24 and any IPv6 link-local addresses.

**Comment by jirib** [comment]
> cupsdNetIFUpdate

```
# grep cupsdNetIFUpdate /var/log/cups/error_log 
D [15/Apr/2021:18:50:06 +0200] cupsdNetIFUpdate: Ignoring "lo".
D [15/Apr/2021:18:50:06 +0200] cupsdNetIFUpdate: Ignoring "enp2s0".
D [15/Apr/2021:18:50:06 +0200] cupsdNetIFUpdate: Ignoring "enp3s0".
D [15/Apr/2021:18:50:06 +0200] cupsdNetIFUpdate: Ignoring "ppp0".
D [15/Apr/2021:18:50:06 +0200] cupsdNetIFUpdate: "lo" = localhost:631
D [15/Apr/2021:18:50:06 +0200] cupsdNetIFUpdate: "enp2s0" = 192.168.1.1:631
D [15/Apr/2021:18:50:06 +0200] cupsdNetIFUpdate: "ppp0" = 88.208.81.140:631
D [15/Apr/2021:18:50:06 +0200] cupsdNetIFUpdate: "lo" = localhost:631
D [15/Apr/2021:18:50:06 +0200] cupsdNetIFUpdate: "enp2s0" = [v1.fe80::225:90ff:fe60:8f1e+enp2s0]:631
D [15/Apr/2021:18:50:06 +0200] cupsdNetIFUpdate: "ppp0" = [v1.fe80::7037:a6be:4f32:a66f+ppp0]:631
```

**Comment by jirib** [comment]
> @jirib I should also note: `@IF(name)` doesn't detect whether the network traffic came over the named interface. Rather, it expands to allow/deny access to all of that interface's subnets. So in your case `@IF(name)` expands to 192.168.1.0/24 and any IPv6 link-local addresses.

The client was having IPv4 192.168.1.194, thus traffic was incoming on *enp2s0* iface on cups server.

**Comment by michaelrsweet** [comment]
@jirib OK, so that all looks correct. What is the address of a client that isn't able to print?

**Comment by michaelrsweet** [comment]
@jirib Never mind, I guess our messages passed each other! :)

If you change the allow line to `Allow 192.168.1.0/24`, does the problem go away?

**Comment by jirib** [comment]
> @jirib OK, so that all looks correct. What is the address of a client that isn't able to print?

I was expecting client to get following output (this one was working as `cupsd.conf` on remote cups server was using `Allow 192.168.1.*`), instead it got `lpstat: Forbidden`.

```
# lpstat -h 192.168.1.1:631 -v
device for epsonalm2000: usb://EPSON/AL-M2000?serial=55PKAZZ21175100705
```

**Comment by jirib** [comment]
> @jirib Never mind, I guess our messages passed each other! :)
> 
> If you change the allow line to `Allow 192.168.1.0/24`, does the problem go away?

Yes. Still the question was

> Is @IF(enp2s0) wrong or it's a bug?

**Comment by michaelrsweet** [comment]
No, `@IF(name)` is expected to work, so this looks like a bug.

**Comment by michaelrsweet** [comment]
Looks like @salgernon found a fix for this:

```
diff --git a/cups/scheduler/auth.c b/cups/scheduler/auth.c
index 4fbad6e2..4fdcf117 100644
--- a/cups/scheduler/auth.c
+++ b/cups/scheduler/auth.c
@@ -202,7 +202,6 @@ cupsdAddNameMask(cups_array_t **masks,      /* IO - Masks array (created as needed) *
 
     if (ifptr >= ifname && *ifptr == ')')
     {
-      ifptr --;
       *ifptr = '\0';
     }
```

I'll be pushing this change later today...

**Comment by michaelrsweet** [comment]
Fixed:

[master d71b7c830] Fix @IF(name) handling (Issue #5918)

OpenPrinting CUPS:

[master 20e1c714b] Fix @IF(name) handling (Apple #5918)
